### PR TITLE
Secure RabbitMQ defaults and authenticate Neutron,Contrail MQ logins

### PR DESCRIPTION
**Pull request by alokjani**
Contrail & Neutron currently connect to MQ using insecure `guest` defaults.
This patch configures rabbitmq w/ a new non-default user using which
contrail and other applications can use to connect.
Also the default `guest` user is removed. This will resolve issue #406.

Note: this patch depends on JioCloud/jiocloud-contrail#31

### CODE REVIEW

File: manifests/rabbitmq.pp [hunk 1]
```diff
@@ -18,6 +18,13 @@
...
   rjil::test { 'check_rabbitmq.sh': }
 
-  include ::rabbitmq
+  #  include ::rabbitmq
+  class { '::rabbitmq':
+    delete_guest_user => true,
+    port              => '5672',
+  }
+
+  ensure_resource('rabbitmq_user', 'rabbit', { password => 'rabbit' })
```
Comment (bodepd) [inline manifests/rabbitmq.pp]: please make this settable from hiera We definitely don't want to swap one hard-coded user/password for another

File: spec/classes/rabbitmq_spec.rb [hunk 2]
```diff
@@ -23,7 +23,8 @@
       should contain_file('/usr/lib/jiocloud/tests/check_rabbitmq.sh')
       should contain_class('rabbitmq').with({
         'manage_repos' => false,
-        'admin_enable' => false
+        'admin_enable' => false,
+				'delete_guest_user' => true,
```
Comment (bodepd) [inline spec/classes/rabbitmq_spec.rb]: formatting issue

**Comment by bodepd** [comment]
you will also need to update this to point to your branch of contrail for it to work.

File: hiera/data/common.yaml [hunk 3]
```diff
@@ -560,13 +560,16 @@ rjil::system::apt::repositories:
 rabbitmq::manage_repos: false
 rabbitmq::admin_enable: false
 
+contrail_rabbit_user: rabbit
+contrail_rabbit_password: rabbit
```
Comment (bodepd) [inline hiera/data/common.yaml]: this should be in secrets/common.yaml not in common

File: manifests/rabbitmq.pp (hunk 1 above)
Comment (alokjani) [inline manifests/rabbitmq.pp]: Sure, will ensure this is settable via hiera.

File: spec/classes/rabbitmq_spec.rb (hunk 2 above)
Comment (alokjani) [inline spec/classes/rabbitmq_spec.rb]: Fixed.

File: hiera/data/common.yaml (hunk 3 above)
Comment (alokjani) [inline hiera/data/common.yaml]: Moved into secrets/common

**Comment by alokjani** [comment]
please test this

**Comment by alokjani** [comment]
please test this

**Comment by alokjani** [comment]
please test this

File: hiera/data/common.yaml [hunk 4]
```diff
@@ -199,7 +199,8 @@ neutron::server::auth_password: "%{hiera('neutron::keystone::auth::password')}"
 neutron::server::auth_host: "%{hiera('keystone_public_address')}"
 neutron::server::auth_protocol: "%{hiera('rjil::keystone::protocol')}"
 neutron::server::mysql_module: 2.3
-neutron::rabbit_password: guest
+neutron::rabbit_user: "%{hiera('neutron_rabbit_user')}"
```
Comment (bodepd) [inline hiera/data/common.yaml]: I thought we were using zeromq for all openstack components @sorenh is this not true? I just noticed that the zeromq definition does not include neutron.

**Comment by alokjani** [comment]
please test this

File: manifests/rabbitmq.pp [hunk 5]
```diff
@@ -20,4 +20,8 @@
 
   include ::rabbitmq
 
+  ensure_resource('rabbitmq_user', "$contrail_rabbit_user", { password => "$contrail_rabbit_password" })
```
Comment (bodepd) [inline manifests/rabbitmq.pp]: why are you using ensure resource here and not just a static resource declaration?

Comment (alokjani) [inline manifests/rabbitmq.pp]: No particular reason. I guess a static resource would be better (and faster).

Comment (bodepd) [inline manifests/rabbitmq.pp]: in general things like create_resources/ensure_resources should only be used when they provide a feature that you need. enusre_resources should be used sparingly and only when you are protecting against a conflict.

**Comment by alokjani** [comment]
please test this

**Comment by alokjani** [comment]
please test this

**Comment by alokjani** [comment]
please test this

**Comment by alokjani** [comment]
please test this

**Comment by alokjani** [comment]
please test this

**Comment by alokjani** [comment]
please test this

**Comment by alokjani** [comment]
please test this

**Comment by alokjani** [comment]
please test this

**Comment by alokjani** [comment]
In build #883 
- found that all neutron & contrail checks has passed and registration w/ consul was successful
- verified that neutron and contrail processes were connected to rabbitmq
- was able to pull down the net-list via neutron indicating that messaging is working 

Somehow, only the stmon checks had failed which shouldn't have as we are in no way touching any stmon related code.

...
Failures:
  Node: stmon2-testjenkins-puppet-rjil-gate-883, Check: validation
  Node: stmon1-testjenkins-puppet-rjil-gate-883, Check: validation
  Node: stmonleader1-testjenkins-puppet-rjil-gate-883, Check: validation
Registered hosts in consul with key name Running_Versions are:
   bootstrap1-testjenkins-puppet-rjil-gate-883
   haproxy1-testjenkins-puppet-rjil-gate-883
   cp1-testjenkins-puppet-rjil-gate-883
   ocdb1-testjenkins-puppet-rjil-gate-883
   oc1-testjenkins-puppet-rjil-gate-883
   ct1-testjenkins-puppet-rjil-gate-883
   cp2-testjenkins-puppet-rjil-gate-883
   gcp1-testjenkins-puppet-rjil-gate-883
...

I'm guessing that these are temporary & build env related so will re-test just to make sure.

**Comment by alokjani** [comment]
please test this

**Comment by alokjani** [comment]
please test this

**Comment by alokjani** [comment]
This is tested this w/ build #930 and is ready for review.  
Note: it depends on  JioCloud/jiocloud-contrail#31

**Comment by bodepd** [comment]
everything is kind of blocked at the moment b/c of some underlying failures

**Comment by bodepd** [comment]
I need this for the kilo upgrade work that I am doing. I'm going to go ahead and take this patch and rebase it myself.

File: Puppetfile [hunk 6]
```diff
@@ -142,8 +142,8 @@ mod 'stephenrjohnson/puppet',
   :ref => '1.0.0'
 
 mod 'jiocloud/contrail',
-  :git => "#{base_url}/jiocloud/jiocloud-contrail",
-  :ref => 'origin/master'
+  :git => "#{base_url}/alokjani/jiocloud-contrail",
+  :ref => 'origin/rabbitmq_auth'
```
Comment (bodepd) [inline Puppetfile]: this is merged.

File: hiera/data/common.yaml [hunk 7]
```diff
@@ -204,7 +210,8 @@ neutron::server::auth_password: "%{hiera('neutron::keystone::auth::password')}"
...
```
Comment (bodepd) [inline hiera/data/common.yaml]: does neutron need rabbitmq? I thought it was using zeromq?

File: hiera/data/secrets/common.yaml [hunk 8]
```diff
@@ -50,6 +50,18 @@ ironic_db_password: ironic
 neutron_db_user: neutron
 neutron_db_password: neutron
 
+# Rabbit MQ
+rabbit_admin_user: rabbit
```
Comment (bodepd) [inline hiera/data/secrets/common.yaml]: why would we not just use the same variable to set all of these?

**Comment by bodepd** [comment]
rebased/updated as #576

File: hiera/data/common.yaml (hunk 7 above)
Comment (alokjani) [inline hiera/data/common.yaml]: Yep, neutron is setup with rabbitmq currently (rpc.impl_kombu). Nova & Cinder are w/ zeromq though.

File: hiera/data/secrets/common.yaml (hunk 8 above)
Comment (alokjani) [inline hiera/data/secrets/common.yaml]: Sure, a hiera lookup of rabbit_admin_user would work as well.

**Comment by alokjani** [comment]
Thanks!  looking forward to this making master via #576

**Pull request closed without merging**